lib/Service/SocialApiService.php

```diff
@@ -290,25 +324,55 @@ public function updateAddressbooks(string $network, string $userId) : JSONRespon
 				continue;
 			}
 
+			// in case this is a follow-up, jump to the last stopped address book
+			if (!is_null($offsetBook)) {
+				if ($addressBook->getURI() !== $offsetBook) {
+					continue;
+				}
+				$offsetBook = null;
+			}
+
 			// get contacts in that addressbook
+			//TODO: activate this optimization when nextcloud/server#22085 is merged
+			/*
+			if (Util::getVersion()[0] < 21) {
+				//TODO: remove this branch when dependency for contacts is min NCv21 (see info.xml)
+				$contacts = $addressBook->search('', ['UID'], ['types' => true]);
+			} else {
+				$contacts = $addressBook->search('', ['X-SOCIALPROFILE'], ['types' => true]);
+			}
+			*/
 			$contacts = $addressBook->search('', ['UID'], ['types' => true]);
-			// TODO: can be optimized by:
-			// $contacts = $addressBook->search('', ['X-SOCIALPROFILE'], ['types' => true]);
-			// but see https://github.com/nextcloud/contacts/pull/1722#discussion_r463782429
-			// and the referenced PR before activating this (index has to be re-created!)
+			usort($contacts, [$this, 'sortContacts']); // make sure the order stays the same in consecutive calls
 
 			// update one contact after another
 			foreach ($contacts as $contact) {
-				// delay to prevent rate limiting issues
-				// TODO: do we need to send an Http::STATUS_PROCESSING ?
-				sleep($delay);
+				// in case this is a follow-up, jump to the last stopped contact
+				if (!is_null($offsetContact)) {
+					if ($contact['UID'] !== $offsetContact) {
+						continue;
+					}
+					$offsetContact = null;
+				}
 
 				try {
 					$r = $this->updateContact($addressBook->getURI(), $contact['UID'], $network);
 					$response = $this->registerUpdateResult($response, $contact['FN'], $r->getStatus());
 				} catch (Exception $e) {
 					$response = $this->registerUpdateResult($response, $contact['FN'], '-1');
 				}
+
+				// stop after 15sec (to be continued with next chunk)
+				if (($this->timeFactory->getTime() - $startTime) > 15) {
+					$response['stoppedAt'] = [
+						'addressBook' => $addressBook->getURI(),
+						'contact' => $contact['UID'],
+					];
+					return new JSONResponse([$response], Http::STATUS_PARTIAL_CONTENT);
+				}
+
+				// delay to prevent rate limiting issues
+				sleep($delay);
```

Comment:
This feels weird. So you sleep here but execute at most 15seconds?

Reply:
Well, this delay is to be compliant to the terms of service of social networks, so I don't want to remove it. And the 15seconds of maximum execution time are to prevent that the update runs forever on huge installations.
How would you propose to tackle this?